### PR TITLE
Add support for blockstorage volume_type quota

### PR DESCRIPTION
**Pull request by nikParasyr**
For #2072 

Api docs:
- https://docs.openstack.org/api-ref/block-storage/v3/?expanded=update-quotas-for-a-project-detail#quota-sets-extension-os-quota-sets

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:
- https://github.com/openstack/cinder/blob/0fe910f130161edf5025d59020093ae163dfcba5/cinder/quota.py#L1157
- https://github.com/openstack/cinder/blob/0fe910f130161edf5025d59020093ae163dfcba5/cinder/quota.py#L1242
- https://github.com/openstack/cinder/blob/0fe910f130161edf5025d59020093ae163dfcba5/cinder/api/contrib/quotas.py#L31
- https://github.com/openstack/cinder/blob/0fe910f130161edf5025d59020093ae163dfcba5/cinder/api/contrib/quotas.py#L192-L275

### CODE REVIEW

**Comment by jtopjian** [comment]
recheck

**Comment by jtopjian** [comment]
@nikParasyr Thanks for working on this. I haven't looked it over since it's still a draft, but wanted to mention that the acceptance test failure is an actual failure. You can see details about the failure by clicking on the `gophercloud-acceptance-test ` link above, and then going to the `job-output.txt.gz` file and searching for:

```
--- FAIL: TestQuotasetUpdate
```

Let me know if you have any questions or need any help. One common issue seen is that people will test on their own local OpenStack deployment which might act differently than a "vanilla" devstack environment. If this is your situation, it might be helpful to launch a vm, install devstack and go and work on the test from within that environment.

**Comment by nikParasyr** [comment]
Hello @jtopjian. Thank you for your input. I'd like some more input and help if possible. The acceptance test fails for 2 reasons. I was aware of both of them from running it locally but i wanted to see the openlab result as well as get an opinion.
- Currently I have updated `TestQuotasetUpdate` to create a volume type foo in order to test that the patch works as intended. That isn't rly workable though because `Extra` is populated by volumes_/snapshots_/gigabytes_<volume_type_name> for every volumeType that exists on the testing cloud. So it doesn't only contain for the foo volume type but also for all existing volumeTypes. The volumeTypes differ from cloud to cloud. So hardcoding the expected Extra map won't work. Therefore to be able to test this properly we probably need to create `Extra` dynamically based on the existing volumeTypes found. The will sort of increase the complexity of the acceptance test but it's the only way i can think of to make it run properly on any testing cloud. Let me know if i should work towards that direction.
- Secondly, even when i hardcoded the Extra based on my testing cloud volume types I still get the following error which I'm a bit unsure how to handle. Same error occurs on openlab. It's a type error. Even more "weird" is that between runs I get different results. One run only gigabytes_foo is complaining, then all of them, then only volumes_foo etc. I'm probably missing something very basic here...
```
	=== RUN   TestQuotasetUpdate
	TestQuotasetUpdate: convenience.go:236: Failure in quotaset_test.go, line 122:at .Extra[gigabytes_foo] expected 100, but got 100
	TestQuotasetUpdate: convenience.go:236: Failure in quotaset_test.go, line 122:at .Extra[volumes_foo] expected 10, but got 10
	TestQuotasetUpdate: convenience.go:236: Failure in quotaset_test.go, line 122:at .Extra[snapshots_foo] expected 10, but got 10
	TestQuotasetUpdate: convenience.go:236: Failure in quotaset_test.go, line 122:at .Extra expected <nil>, but got map[string]interface {}{"gigabytes_foo":100, "snapshots_foo":10, "volumes_foo":10}
	TestQuotasetUpdate: convenience.go:35: Failure in quotaset_test.go, line 122: The structures were different.
	--- FAIL: TestQuotasetUpdate (0.88s)
	FAIL
```

**Comment by jtopjian** [comment]
> Currently I have updated TestQuotasetUpdate to create a volume type foo in order to test that the patch works as intended. That isn't rly workable though because Extra is populated by volumes_/snapshots_/gigabytes_<volume_type_name> for every volumeType that exists on the testing cloud.

This is OK. Take a look at the other acceptance tests for different examples. A lot of them contain a pattern that defines a variable called `found` and then only checks for a specific resource, ignoring other resources. This is to account for existing resources in the cloud.

> Secondly, even when i hardcoded the Extra based on my testing cloud volume types I still get the following error which I'm a bit unsure how to handle. 

At first glance, this looks like some kind of type assertion issue. I would double-check the type that is being marshaled into "Extra" and make sure it's being compared to a type of the same kind.

I hope this helps, but let me know if you need more details.

**Comment by nikParasyr** [comment]
@jtopjian Thanks for all the help. This is ready for a first review.
Tried to follow the contributing guide as much as possible but I might have missed some spots.

**Comment by nikParasyr** [comment]
recheck

**Comment by nikParasyr** [comment]
recheck

**Comment by nikParasyr** [comment]
recheck

**Comment by jtopjian** [comment]
@nikParasyr There might be an issue with OpenLab. I recommend trying again in a few hours. I'll try to get some time to review this either today or tomorrow.

**Comment by nikParasyr** [comment]
recheck

File: openstack/blockstorage/extensions/quotasets/results.go [hunk 1]
```diff
@@ -34,6 +36,42 @@ type QuotaSet struct {
...
+	}
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+	*r = QuotaSet(s.tmp)
+
+	// Collect other fields and bundle them into Extra
+	// but only if a field titled "extra" wasn't sent.
+	if s.Extra != nil {
```
Comment (jtopjian) [inline openstack/blockstorage/extensions/quotasets/results.go:61]: This is specific to the Keystone/Identity API and can be removed here. The body of the else clause should only be used.

**Comment by jtopjian** [comment]
recheck

File: openstack/blockstorage/extensions/quotasets/results.go (hunk 1 above)
Comment (nikParasyr) [reply openstack/blockstorage/extensions/quotasets/results.go]: done

**Comment by nikParasyr** [comment]
hmmmm openlab is consistently failing. maybe [this](https://github.com/theopenlab/openlab/issues/674) is related? not the occm stuff but openlab upgrading to focal? ill stop rechecking for the time being

**Comment by jtopjian** [comment]
Yes, it's possible. I've subscribed to the issue and will keep an eye on it.

In the meantime, I have run the tests in my own devstack environment and everything looks good to me. If you're finished with this, I can go ahead and merge it.

**Comment by nikParasyr** [comment]
Yes. I've finished with this one. Thank you for all the help :)

**Review by jtopjian — APPROVED** [review]
LGTM - thank you!

**Comment by jtopjian** [comment]
@nikParasyr Thank you for submitting this - this was really good work.

**Pull request merged**